### PR TITLE
[DataTable] Allow JSX in the headers

### PR DESCRIPTION
**Pull request by devonpmack**
### WHY are these changes introduced?

Fixes https://github.com/Shopify/polaris-react/issues/2597 

We needed check boxes in our header but it was not supported. (Part of https://github.com/Shopify/store/issues/12522)

### WHAT is this pull request doing?

Allow the header prop of DataTable to receive and display React Nodes.

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from "react";
import { Card, DataTable, Page, Button } from "@shopify/polaris";

export default function DataTableExample() {
  return (
    <Page title="Sales by product">
      <Card>
        <DataTable
          columnContentTypes={[Array(5).fill("text")]}
          headings={Array(5).fill(<Button />)}
          rows={[]}
        />
      </Card>
    </Page>
  );
}


```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [x] Updated the component's `README.md` with documentation changes
* [x] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the Polaris UI kit

### CODE REVIEW

**Comment by devonpmack** [comment]
@PabloVallejo Good call, I just added a test and example to component README.

File: src/components/DataTable/DataTable.tsx [hunk 1]
```diff
@@ -29,7 +29,7 @@ export interface DataTableProps {
   /** List of data types, which determines content alignment for each column. Data types are "text," which aligns left, or "numeric," which aligns right. */
   columnContentTypes: ColumnContentType[];
   /** List of column headings. */
-  headings: string[];
+  headings: (string | React.ReactNode)[];
```
Comment (BPScott) [inline src/components/DataTable/DataTable.tsx]: `React.ReactNode` already contains the type of `string`, so just `React.ReactNode[]` is sufficient

File: src/components/DataTable/README.md [hunk 2]
```diff
@@ -334,7 +334,7 @@ function FullDataTableExample() {
             'numeric',
           ]}
           headings={[
-            'Product',
+            <TextStyle variation="strong">Product</TextStyle>,
```
Comment (dleroux) [inline src/components/DataTable/README.md]: I'm not sure I would update this example. Sort of implies it's a pattern we want to promote.

```suggestion
            'Product',
```

Comment (devonpmack) [reply src/components/DataTable/README.md]: Okay, fair enough

**Pull request merged**